feat(node): stop actor system if any essential actor panics

tmpolaczyk opened the pull request:
Fix #1934

How to test: add a panic somewhere in the node and check that the node stops after showing the panic message.

lrubiorod commented on node/src/actors/rad_manager/actor.rs:
```diff
@@ -10,6 +10,11 @@ impl Actor for RadManager {
+
+    fn stopped(&mut self, _ctx: &mut Self::Context) {
+        // RadManager is expected to restart on panic
+        //System::current().stop_with_code(1);
```
In this case, when the RadManager dies, it looks that it would never recover

tmpolaczyk replied:
I assumed that it is restarted automatically, will try to add tests to see which actors restart and which ones stop the system when they panic.

tmpolaczyk replied:
Currently all actors stop the system on panic, (except `Session` and `JsonRpc`, which handle client connections and disconnect on panic). `RadManager` executes data requests inside a future, and that makes it immune to radon panics. It's not that the `RadManager` restarts automatically, it's that it never stops.

lrubiorod commented on node/src/signature_mngr.rs:
```diff
@@ -257,6 +257,10 @@ impl Actor for SignatureManager {
     }
+
+    fn stopped(&mut self, _ctx: &mut Self::Context) {
+        System::current().stop_with_code(1);
```
I include a panic in Handler<GetPkh> in line 379 and it doesnt close the main node

lrubiorod replied:
I am seeing that it depends in which part of the code panic, in the handlers there would not close the main node, but it will do if the panic is in the functions that send the messages to the handlers

tmpolaczyk replied:
Maybe it's because we also need to add `System::current().stop_with_code` to `StorageManagerAdapter`?

tmpolaczyk replied:
> Maybe it's because we also need to add System::current().stop_with_code to StorageManagerAdapter?

No, that didn't work. But I fixed it by adding an `impl Drop for StorageManager` that detects if the thread is panicking.

tmpolaczyk replied:
This happens because the `SignatureManager` is started using `SyncArbiter::start`. When a `SignatureManager` handler panics, the panic propagates to the `SyncArbiter`, and that kills the actor without calling its `stopped` method. The `SignatureManagerAdapter` stores a reference (`Addr<SignatureManager>`) to this `SyncArbiter` and when the `SyncArbiter` dies, this reference is no longer valid. We could fix it by doing something like:

```rust
if !self.addr.connected() {
    // The StorageManager died, restart
    self.addr = SyncArbiter::start(...);
}
```

But I think that is too complex, and the alternative solution using `impl Drop` is more robust.

tmpolaczyk commented:
This is a rocksdb error:

```
pure virtual method called
terminate called without an active exception
```

we fixed it in the wallet by waiting for the database to close before killing the process, we can do something similar in the node.

tmpolaczyk commented:
> This is a rocksdb error:
> 
> ```
> pure virtual method called
> terminate called without an active exception
> ```
> 
> we fixed it in the wallet by waiting for the database to close before killing the process, we can do something similar in the node.

I added a sleep and it seems to work good enough, we can create an issue to fix it properly

lrubiorod commented on node/src/actors/chain_manager/mod.rs:
```diff
@@ -71,6 +71,7 @@ use witnet_validations::validations::{
     validate_new_transaction, verify_signatures, VrfSlots,
 };
 
+use crate::utils::stop_system_if_panicking;
```
fmt

tmpolaczyk commented on node/src/actors/rad_manager/mod.rs:
```diff
@@ -12,3 +12,11 @@ mod handlers;
+impl Drop for RadManager {
+    fn drop(&mut self) {
+        log::trace!("Dropping RadManager");
+        // TODO: RadManager is expected to restart on panic, ensure
```
It currently does not restart on panic.

tmpolaczyk commented:
Finally I did not add tests because not sure to how ensure that every actor has the expected behavior.
